File: 0312-burst-balloons/0312-burst-balloons.java
class Solution {
    public int maxCoins(int[] nums) {
        int n = nums.length;
        
        // Add padding
        int[] arr = new int[n + 2];
        arr[0] = 1;
        arr[n + 1] = 1;
        for (int i = 0; i < n; i++) {
            arr[i + 1] = nums[i];
        }

        int[][] dp = new int[n + 2][n + 2];

        // length is the size of the range (not actual length of the subarray)
        for (int length = 2; length < arr.length; length++) {
            for (int left = 0; left < arr.length - length; left++) {
                int right = left + length;
                for (int i = left + 1; i < right; i++) {
                    int coins = arr[left] * arr[i] * arr[right];
                    int total = dp[left][i] + coins + dp[i][right];
                    dp[left][right] = Math.max(dp[left][right], total);
                }
            }
        }

        return dp[0][n + 1]; // whole range (excluding virtual balloons)

    }
}

File: 0312-burst-balloons/0312-burst-balloons.swift
class Solution {
    func maxCoins(_ nums: [Int]) -> Int {
        let n = nums.count
        let arr = [1] + nums + [1]  // Padding with 1s
        let size = arr.count
        // Initialize 2D dp array with zeros
        var dp = Array(repeating: Array(repeating: 0, count: size), count: size)
        
        for length in 2..<arr.count{
            for left in 0..<(size-length){
                let right = left + length
                for i in (left+1)..<right{
                    let coins = arr[left] * arr[i] * arr[right]
                    let total = dp[left][i] + coins + dp[i][right]
                    dp[left][right] = max(dp[left][right], total)
                }
            }
        }
        return dp[0][n+1]
    }
}